zookeeper/source/zk_nn_client.cc

```diff
@@ -1322,14 +1317,70 @@ bool ZkNnClient::add_block(const std::string &file_path,
   return true;
 }
 
+bool ZkNnClient::add_block_group(const std::string &fileName,
+                     u_int64_t &block_group_id,
+                     std::vector<std::string> &dataNodes,
+                     std::vector<u_int64_t> &storageBlockIDs,
+                     uint32_t total_num_storage_blocks) {
+
+    FileZNode znode_data;
+    read_file_znode(znode_data, fileName);
+
+    // Generate the block group id and storage block ids.
+    block_group_id = generate_block_group_id();
+    for (u_int64_t i = 0; i < total_num_storage_blocks; i++) {
+        storageBlockIDs.push_back(generate_storage_block_id(block_group_id, i));
+    }
+
+    // Log them for debugging purposes.
+    LOG(INFO) << "Generated block group id " << block_group_id << "\n";
+    for (auto storageBlockID : storageBlockIDs)
+        LOG(INFO) << "Generated storage block id " << storageBlockID << "\n";
+
+    // TODO(nate): figure out what exact zookeepr operations must occur.
+    return true;
+}
+
 
-u_int64_t ZkNnClient::generate_hierarchical_block_id(
-        uint64_t block_group_id,
-        uint32_t index_in_group) {
-    // TODO(Nate): actually implement this naming scheme.
-    return 0;
+u_int64_t ZkNnClient::generate_storage_block_id(
+        u_int64_t block_group_id,
+        u_int64_t index_within_group) {
+    // TODO(nate): assume a file = no more than 2**47 128 MB blocks
+    // TODO(nate): assume no more than 2**16 storage blocks in a logical block
+    u_int64_t res = 1ull << 63;  // filled with 1 and 63 zeros.
+    res = res & block_group_id;  // & to fill bit2~bit48.
+    res = res & index_within_group;  // & to fill the lower 16 bits.
+    return res;
 }
 
+u_int64_t ZkNnClient::generate_block_group_id() {
+    u_int64_t res;
+    util::generate_uuid(res);  // generate some random 64 bits.
+    res = res | (1ull << 63);  // filp the highest bit to one.
+    res = (res >> 16) << 16;  // fill the lower 16 bits with zeros.
+    return res;
+}
+
+u_int64_t ZkNnClient::get_block_group_id(u_int64_t storage_block_id) {
+    u_int64_t mask = ((1ull << 47) - 1) << 16;  // 47 ones and 16 zeros.
+    return storage_block_id & mask;
+}
+
+u_int64_t ZkNnClient::get_index_within_block_group(u_int64_t storage_block_id) {
+    u_int64_t mask = 0xffff;  // 48 zeroes and 16 ones.
+    return storage_block_id & mask;
+}
+
+
+uint32_t ZkNnClient::get_total_num_storage_blocks(
+        const std::string &fileName,
+        u_int64_t &block_group_id) {
+    // TODO(nate): actually figure out where this information is stored.
+    // TODO(nate): is it supposed to be figured out from ecPolicyName?
+    return 9; // arbitrarily assume RS(6, 3)
```

Comment:
I think this is a fine implementation for now. Our first iteration of EC can enforce RS(6, 3) and add the ability to specify later.